core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrepImp.java

```diff
@@ -0,0 +1,208 @@
+        List<String> matchedLines =  new ArrayList<>();;
+
+        for(File file : files){
+            List<String> lines = this.readLines(file);
+            boolean matchFound = false;
+            for(String line: lines){
+                matchFound = this.containsPattern(line);
+
+                //check if match is found
+                if(matchFound){
+                    matchedLines.add(line);
+                }
+            }
+        }
+
+        //check if matched Lines found and write to file
+        if(matchedLines!=null){
+            this.writeToFile(matchedLines);
+            System.out.println("Patterns has been written to File");
+        }else{
+            System.out.println("Pattern Not found");
+        }
+
+    }
+
+    //traverse a directory and return all the files
+    @Override
+    public List<File> listFiles(String rootDir)
+    {
+        File directory = new File(rootDir);
+
+        // Check if rootDir is Directory
+        if(!directory.exists() || !directory.isDirectory()){
+            throw new IllegalArgumentException(String.format("%s is not a directory", rootDir));
+        }
+
+        List<File> fileList = new ArrayList<>();
+        File rootdir = new File(rootDir);
+
+        //Get the name of directories contained in the dir
+        if (rootdir.isDirectory()) {
+            File[] files = rootdir.listFiles();
+
+
+        if (files != null) {
+             for (File file : files) {
+
+                 //check if file, then add to list
+                  if (file.isFile()) {
+                      fileList.add(file);
+                  } else if (file.isDirectory()) {
+
+                      //add files into list from sub-dir recursively
+                      fileList.addAll(listFiles(file.getAbsolutePath()));
+                  }
+             }
+          }
+      }
+        return fileList;
+
+    }
+
+    //Read a file and return all the lines
+    @Override
+    public List<String> readLines(File inputFile)
+    {
+        // Check if inputFile is a file
+        if(!inputFile.exists() || !inputFile.isFile()){
+            throw new IllegalArgumentException(String.format("%s is not a file", inputFile.getPath()));
+        }
+
+        List<String> result = new ArrayList<>();
+        BufferedReader br = null;
+
+        try {
+            br = new BufferedReader(new FileReader(inputFile));
+
+            String line;
+            while ((line = br.readLine()) != null) {
+                result.add(line);
+            }
+            return result;
+
+        } catch (IOException e) {
+             e.printStackTrace();
+        }
+        return null;
+    }
+
+    //check if a line contains the regex pattern (passed by user)
+    @Override
+    public boolean containsPattern(String line)
+    {
+        PatternSyntaxException exc = null;
+        try {
+            Pattern.compile(regex);
+        } catch (PatternSyntaxException e) {
+            exc = e;
```

Comment:
Print your stack trace here.  A method will exit after the catch block is executed so if you do catch an exception you program will never print the stack trace.